Source/PLCrashAsyncMachOImage.c

```diff
@@ -491,64 +536,23 @@ plcrash_error_t plcrash_async_macho_map_segment (plcrash_async_macho_t *image, c
+    if (segment == NULL) {
         return PLCRASH_ENOTFOUND;
-
-    cmd_32 = segment;
-    cmd_64 = segment;
-    
-    uint32_t nsects;
-    uintptr_t cursor = (uintptr_t) segment;
-
-    if (image->m64) {
-        nsects = image->byteorder->swap32(cmd_64->nsects);
-        cursor += sizeof(*cmd_64);
-    } else {
-        nsects = image->byteorder->swap32(cmd_32->nsects);
-        cursor += sizeof(*cmd_32);
     }
-
-    for (uint32_t i = 0; i < nsects; i++) {        
-        struct section *sect_32 = NULL;
-        struct section_64 *sect_64 = NULL;
-       
-        if (image->m64) {
-            if (!plcrash_async_mobject_verify_local_pointer(&image->load_cmds, cursor, 0, sizeof(*sect_64))) {
-                PLCF_DEBUG("Section table entry outside of expected range; searching for (%s,%s)", segname, sectname);
-                return PLCRASH_EINVAL;
-            }
-            
-            sect_64 = (void *) cursor;
-            cursor += sizeof(*sect_64);
-        } else {
-            if (!plcrash_async_mobject_verify_local_pointer(&image->load_cmds, cursor, 0, sizeof(*sect_32))) {
-                PLCF_DEBUG("Section table entry outside of expected range; searching for (%s,%s)", segname, sectname);
-                return PLCRASH_EINVAL;
-            }
-            
-            sect_32 = (void *) cursor;
-            cursor += sizeof(*sect_32);
+    uintptr_t cursor = (uintptr_t) segment;
+    uint32_t nsects = plcrash_async_macho_read_sections_count(image, &cursor);
+    for (uint32_t i = 0; i < nsects; i++) {
+        const char *image_sectname;
+        pl_vm_address_t sectaddr;
+        pl_vm_size_t sectsize;
+        if (!plcrash_async_macho_read_section(image, &cursor, &image_sectname, &sectaddr, &sectsize)) {
+            PLCF_DEBUG("Section table entry outside of expected range; searching for (%s,%s)", segname, sectname);
+            return PLCRASH_EINVAL;
         }
-        
-        const char *image_sectname = image->m64 ? sect_64->sectname : sect_32->sectname;
-        if (plcrash_async_strncmp(sectname, image_sectname, sizeof(sect_64->sectname)) == 0) {
-            /* Calculate the in-memory address and size */
-            pl_vm_address_t sectaddr;
-            pl_vm_size_t sectsize;
-            if (image->m64) {
-                sectaddr = image->byteorder->swap64(sect_64->addr) + image->vmaddr_slide;
-                sectsize = image->byteorder->swap64(sect_64->size);
-            } else {
-                sectaddr = image->byteorder->swap32(sect_32->addr) + image->vmaddr_slide;
-                sectsize = image->byteorder->swap32(sect_32->size);
-            }
-            
-            
+        if (plcrash_async_strncmp(sectname, image_sectname, fldsiz(section_64, sectname)) == 0) {
             /* Perform and return the mapping */
+            // PLCF_DEBUG("%s (%s,%.*s): 0x%lx - 0x%lx", PLCF_DEBUG_IMAGE_NAME(image), segname, (int)fldsiz(section_64, sectname), image_sectname, sectaddr, sectaddr + sectsize);
```

Comment:
Is this still useful?

Reply:
Yes, I'd keep it for future debugging. It can be useful when you cannot attach real debugger